### PR TITLE
Add validation extension to script

### PR DESCRIPTION
**Pull request by tcharding**
Add an extension trait for the validation logic in preparation for moving the `Script` type to `primitives`.

### CODE REVIEW

**Comment by tcharding** [comment]
If this merges we should do the same to the `Transaction` impl block.

File: bitcoin/src/consensus/validation.rs [hunk 1]
```diff
@@ -146,12 +146,12 @@ impl Script {
...
     ///
     /// [`bitcoinconsensus::VERIFY_ALL_PRE_TAPROOT`]: https://docs.rs/bitcoinconsensus/0.106.0+26.0/bitcoinconsensus/constant.VERIFY_ALL_PRE_TAPROOT.html
-    pub fn verify_with_flags<F: Into<u32>>(
+    pub fn verify_with_flags(
         &self,
         index: usize,
         amount: crate::Amount,
         spending_tx: &[u8],
-        flags: F,
+        flags: impl Into<u32>,
```
Comment (Kixunil) [inline bitcoin/src/consensus/validation.rs]: This is strictly a breaking change. I don't even know why there's this `Into<u32>` and why not have just `u32` or even better a newtype representing flags.

But I don't mind this change.

**Review by Kixunil — APPROVED** [review]
ACK 70b282f091c057b2f438953a4bf161ad2d6927c1

File: bitcoin/src/consensus/validation.rs (hunk 1 above)
Comment (tcharding) [reply bitcoin/src/consensus/validation.rs]: Oh it is? I thought I read that this was syntax sugar for the exact same code it changes?

Comment (tcharding) [reply bitcoin/src/consensus/validation.rs]: Oh I see, one cannot use turbo fish now after this change so you are correct its a breaking change. I'll update the git log.

**Comment by tcharding** [comment]
This PR displays the new API breaking bot not giving accurate coverage. Patch 1 is subtly a breaking change, and missed by the bot. More concerning, the second patch is obviously and always was an  API breaking change.

**Comment by tcharding** [comment]
Force push is update to commit log only, no other changes.

**Comment by Kixunil** [comment]
How come the update didn't dismiss my ACK though? Some new GH update?

**Comment by tcharding** [comment]
> Nobody claimed it's perfect. 🤷‍♂️

Just out of interest, I had in mind when we introduced it that I would filter by label to do the release notes, were you expecting the bots labels to be good enough for that?

**Comment by Kixunil** [comment]
> were you expecting the bots labels to be good enough for that?

I was expecting the bot to see a problem in a simple case like this (but maybe it's not really simple to implement) but generally I expected that if neither we nor the bot notice a break we can't do any better anyway and we can add the label whenever we do notice and bot doesn't (we can also remove it in case of FPs.)

File: bitcoin/src/consensus/validation.rs (hunk 1 above)
Comment (apoelstra) [reply bitcoin/src/consensus/validation.rs]: oo, that's subtle. I also missed that this was breaking.

**Comment by apoelstra** [comment]
Needs rebase.

**Comment by Kixunil** [comment]
Crap, every time I read the title I think this adds an interpreter... (The horror :scream: )

**Comment by apoelstra** [comment]
Needs rebase again. Sorry, I've been active today merging all the "easy" stuff in my notifications.

**Review by Kixunil — APPROVED** [review]
ACK 298b96c5792f19602e05acfe56e8d1f121648677

**Review by apoelstra — APPROVED** [review]
ACK 298b96c5792f19602e05acfe56e8d1f121648677

**Pull request merged**